File: tests/cpu/seats/city-rows.test.ts
import { describe, it, expect } from 'vitest';
import { makeMap, makeState, tileAtCoords, settleAt, grantTechs, grantCivics } from '../helpers';
import { emptySeat, setTileOwner } from '../../../cpu/core/seats';
import { computeCityStats } from '../../../cpu/core/city';
import { getModifiers, prodMultFor } from '../../../cpu/core/effects';
import { greatPersonPointsPerTurn } from '../../../cpu/core/greatPeople';
import { poweredExtra } from '../../../cpu/core/yields';
import { accrueStockpiles, stockpileCap } from '../../../cpu/core/stockpile';
import { extraCharges, spawnUnit } from '../../../cpu/core/units';
import { tilePurchaseCost, foundCityAt } from '../../../cpu/core/game';
import { validImprovements } from '../../../cpu/core/rules';
import { cityTradeYields } from '../../../cpu/core/trade';
import { spyCapacity } from '../../../cpu/core/espionage';
import { seatPhase } from '../../../cpu/core/phase';
import { neighbors } from '../../../world/hex';
import { CIV_LEADERS } from '../../../cpu/data/seats';
import { PROD_MULT_ROWS, CENTER_ADJ_ROWS, GREAT_WORK_YIELD_ROWS, POWERED_YIELD_ROWS, STOCKPILE_RATE_ROWS, GRANT_UNIT_ROWS, CAPITAL_ROWS } from '../../../cpu/data/civilizations';
import { STRATEGIC_IDS, STOCKPILE_CAP_BASE } from '../../../cpu/data/constants';
import { TECHS } from '../../../cpu/data/techs';
import type { GameState } from '../../../cpu/core/types';

/**
 * THE CITY'S ROSTER ROWS (CIV6, the install's TraitModifiers): the centre's
 * terrain adjacency, the per-work yields, the Great Person factor, the
 * powered building's extra yield, strategic accumulation and its ceiling,
 * build charges, the tile price, the Farm's ground, the route's per-
 * improvement yields, the granted units, spy capacity and the first city.
 */
const seatRow = (civ: string) => CIV_LEADERS.findIndex((l) => l.civ === civ);
const leaderRow = (leader: string) => CIV_LEADERS.findIndex((l) => l.leader === leader);

function sceneAs(row: number, terrain: 'GRASSLAND' | 'DESERT' | 'TUNDRA' = 'GRASSLAND'): GameState {
  const state = makeState(makeMap(14, 14, terrain));
  state.seats.push(emptySeat(1));
  state.seats[0].civ = row;
  state.seats[1].civ = seatRow('AMERICA');
  return state;
}

describe('the production rows widen', () => {
  it('name a district item, one unit type, or every item of a kind', () => {
    expect(PROD_MULT_ROWS.length).toBe(13);
    const hojo = PROD_MULT_ROWS.filter((r) => r.leader === 'HOJO');
    expect(prodMultFor(hojo, { kind: 'district', districtItem: 'ENCAMPMENT' })).toBe(2);
    expect(prodMultFor(hojo, { kind: 'district', districtItem: 'CAMPUS' })).toBe(1);
    const dutch = PROD_MULT_ROWS.filter((r) => r.civ === 'NETHERLANDS');
    expect(prodMultFor(dutch, { kind: 'district', districtItem: 'DAM' })).toBe(1.5);
    const mali = PROD_MULT_ROWS.filter((r) => r.civ === 'MALI');
    expect(prodMultFor(mali, { kind: 'building', building: 'MONUMENT' })).toBe(0.7);
    expect(prodMultFor(mali, { kind: 'unit', unit: 'WARRIOR', promoClass: 'MELEE' })).toBe(0.7);
    expect(prodMultFor(mali, { kind: 'district', districtItem: 'CAMPUS' })).toBe(1);
    const eng = PROD_MULT_ROWS.filter((r) => r.civ === 'ENGLAND');
    expect(prodMultFor(eng, { kind: 'unit', unit: 'MILITARY_ENGINEER', promoClass: 'SUPPORT' })).toBe(2 * 1);
    expect(prodMultFor(eng, { kind: 'unit', unit: 'WARRIOR', promoClass: 'MELEE' })).toBe(1);
  });

  it("Grote Rivieren: the Netherlands' three districts read the river as a major adjacency", () => {
    const state = sceneAs(seatRow('NETHERLANDS'));
    const add = getModifiers(state, 0).districtAdjacencyAdd;
    expect(add.CAMPUS?.[0]).toEqual({ source: 'RIVER', amount: 2 });
    expect(add.INDUSTRIAL_ZONE?.[0].source).toBe('RIVER');
    expect(getModifiers(state, 1).districtAdjacencyAdd.CAMPUS ?? []).toEqual([]);
  });
});

describe("Songs of the Jeli", () => {
  it("pays Mali's centre per adjacent Desert tile, and nobody else's", () => {
    expect(CENTER_ADJ_ROWS.length).toBe(2);
    const faithOf = (row: number): number => {
      const state = sceneAs(row, 'DESERT');
      const city = settleAt(state, tileAtCoords(state.map, 7, 7).index, 0);
      return computeCityStats(state, city).breakdown.tiles.faith;
    };
    const centre = tileAtCoords(makeMap(14, 14, 'DESERT'), 7, 7);
    const n = neighbors(makeMap(14, 14, 'DESERT'), centre).length;
    expect(faithOf(seatRow('MALI'))).toBe(faithOf(seatRow('AMERICA')) + n);
  });
});

describe('Nkisi', () => {
  it("pays Kongo per Relic and Artifact, and 50% more points for three classes", () => {
    expect(GREAT_WORK_YIELD_ROWS.length).toBe(8);
    const state = sceneAs(seatRow('KONGO'));
    const city = settleAt(state, tileAtCoords(state.map, 7, 7).index, 0);
    city.relics = 2;
    city.artifacts = 1;
    const kongo = computeCityStats(state, city).breakdown.buildings;
    const plain = sceneAs(seatRow('AMERICA'));
    const pc = settleAt(plain, tileAtCoords(plain.map, 7, 7).index, 0);
    pc.relics = 2;
    pc.artifacts = 1;
    const none = computeCityStats(plain, pc).breakdown.buildings;
    expect(kongo.gold - none.gold).toBe(4 * 3);
    expect(kongo.food - none.food).toBe(2 * 3);
    expect(kongo.production - none.production).toBe(2 * 3);
    expect(kongo.faith - none.faith).toBe(1 * 3);
    expect(getModifiers(state, 0).gppClassMult.ARTIST).toBe(1.5);
    expect(getModifiers(state, 0).gppClassMult.SCIENTIST).toBeUndefined();
    // the factor rides every per-turn source
    const pts = greatPersonPointsPerTurn(state, 0);
    expect(pts.ARTIST).toBe(0); // no Theater Square yet — the factor over zero
  });
});

describe('Workshop of the World', () => {
  it("adds +4 to each yield a POWERED building pays, England alone", () => {
    expect(POWERED_YIELD_ROWS.length).toBe(5);
    const state = sceneAs(seatRow('ENGLAND'));
    settleAt(state, tileAtCoords(state.map, 7, 7).index, 0);
    expect(poweredExtra(getModifiers(state, 0), { science: 5 })).toEqual({ science: 4 });
    expect(poweredExtra(getModifiers(state, 0), { culture: 2, science: 0 })).toEqual({ culture: 4 });
    expect(poweredExtra(getModifiers(state, 1), { science: 5 })).toEqual({});
  });

  it('accumulates 2 more Iron and Coal per mine, and 10 more capacity per harbour building', () => {
    expect(STOCKPILE_RATE_ROWS.length).toBe(4);
    const bankOf = (row: number, resource: 'IRON' | 'COAL'): number => {
      const state = sceneAs(row);
      const city = settleAt(state, tileAtCoords(state.map, 7, 7).index, 0);
      const t = tileAtCoords(state.map, 7, 8);
      setTileOwner(t, 0, city.id);
      t.resource = resource;
      t.improvement = 'MINE';
      accrueStockpiles(state, 0);
      return state.seats[0].stockpile![STRATEGIC_IDS.indexOf(resource)];
    };
    expect(bankOf(seatRow('ENGLAND'), 'IRON')).toBe(bankOf(seatRow('AMERICA'), 'IRON') + 2);
    expect(bankOf(seatRow('ENGLAND'), 'COAL')).toBe(bankOf(seatRow('AMERICA'), 'COAL') + 2);
    const state = sceneAs(seatRow('ENGLAND'));
    const city = settleAt(state, tileAtCoords(state.map, 7, 7).index, 0);
    expect(stockpileCap(state, 0)).toBe(STOCKPILE_CAP_BASE);
    city.buildings.push('LIGHTHOUSE');
    expect(stockpileCap(state, 0)).toBe(STOCKPILE_CAP_BASE + 10);
    city.buildings.push('SEAPORT');
    expect(stockpileCap(state, 0)).toBe(STOCKPILE_CAP_BASE + 20);
  });

  it('gives a Military Engineer two more charges', () => {
    const eng = sceneAs(seatRow('ENGLAND'));
    const at = tileAtCoords(eng.map, 7, 7);
    expect(extraCharges(eng, 0, 'MILITARY_ENGINEER', at)).toBe(2);
    expect(extraCharges(eng, 0, 'BUILDER', at)).toBe(0);
    expect(extraCharges(eng, 1, 'MILITARY_ENGINEER', at)).toBe(0);
  });
});

describe('The Last Best West', () => {
  it("halves Canada's tundra tile price and farms the tundra, hills at Civil Engineering", () => {
    const price = (row: number): number => {
      const state = sceneAs(row, 'TUNDRA');
      const city = settleAt(state, tileAtCoords(state.map, 7, 7).index, 0);
      return tilePurchaseCost(state, city, tileAtCoords(state.map, 7, 9).index);
    };
    expect(price(leaderRow('LAURIER'))).toBe(Math.round(price(seatRow('AMERICA')) * 0.5));
    const state = sceneAs(leaderRow('LAURIER'), 'TUNDRA');
    settleAt(state, tileAtCoords(state.map, 7, 7).index, 0);
    grantTechs(state, 'POTTERY', 'IRRIGATION');
    const flat = tileAtCoords(state.map, 7, 8);
    setTileOwner(flat, 0, state.seats[0].cities[0].id);
    expect(validImprovements(state, flat, 0)).toContain('FARM');
    const hill = tileAtCoords(state.map, 7, 9);
    hill.elevation = 'HILLS';
    setTileOwner(hill, 0, state.seats[0].cities[0].id);
    expect(validImprovements(state, hill, 0)).not.toContain('FARM');
    grantCivics(state, 'CIVIL_ENGINEERING');
    expect(validImprovements(state, hill, 0)).toContain('FARM');
    // a plain seat farms neither
    const plain = sceneAs(seatRow('AMERICA'), 'TUNDRA');
    settleAt(plain, tileAtCoords(plain.map, 7, 7).index, 0);
    grantTechs(plain, 'POTTERY', 'IRRIGATION');
    const pf = tileAtCoords(plain.map, 7, 8);
    setTileOwner(pf, 0, plain.seats[0].cities[0].id);
    expect(validImprovements(plain, pf, 0)).not.toContain('FARM');
  });
});

describe('Favorable Terms', () => {
  it("pays Poundmaker per Camp at the route's destination, and per Camp here on a route in", () => {
    const state = sceneAs(leaderRow('POUNDMAKER'));
    const from = settleAt(state, tileAtCoords(state.map, 3, 3).index, 0);
    const to = settleAt(state, tileAtCoords(state.map, 9, 9).index, 0);
    const camp = tileAtCoords(state.map, 9, 8);
    setTileOwner(camp, 0, to.id);
    camp.resource = 'DEER';
    camp.improvement = 'CAMP';
    state.seats[0].tradeRoutes = [{ from: from.id, to: to.id, turnsLeft: 20 } as never];
    const out = cityTradeYields(state, from, 0);
    expect(out.food).toBe(1 + 1); // routeYields' own food, plus the Camp
    // the destination side: another seat's route in pays this seat gold per Camp here
    const homeCamp = tileAtCoords(state.map, 3, 4);
    setTileOwner(homeCamp, 0, from.id);
    homeCamp.resource = 'DEER';
    homeCamp.improvement = 'CAMP';
    state.seats[1].tradeRoutes = [{ from: 1, toSeat: 0, toSeatCity: from.id, turnsLeft: 20 } as never];
    // one foreign route in pays the origin per Camp HERE; the domestic route
    // this seat sent pays its own destination the same way
    expect(cityTradeYields(state, from, 0).gold).toBe(1);
    expect(cityTradeYields(state, to, 0).gold).toBe(1);
    const plain = sceneAs(seatRow('AMERICA'));
    const pf = settleAt(plain, tileAtCoords(plain.map, 3, 3).index, 0);
    const pt = settleAt(plain, tileAtCoords(plain.map, 9, 9).index, 0);
    const pc = tileAtCoords(plain.map, 9, 8);
    setTileOwner(pc, 0, pt.id);
    pc.resource = 'DEER';
    pc.improvement = 'CAMP';
    plain.seats[0].tradeRoutes = [{ from: pf.id, to: pt.id, turnsLeft: 20 } as never];
    expect(cityTradeYields(plain, pf, 0).food).toBe(1);
    expect(cityTradeYields(plain, pt, 0).gold).toBe(0);
  });
});

describe('the granted units', () => {
  it('the Cree Trader at Pottery, and nobody else at that technology', () => {
    // the TECH-keyed rows, not the list's length: the family also carries
    // founding grants now, and a bare count moves with every one of them
    expect(GRANT_UNIT_ROWS.filter((r) => r.tech !== undefined).length).toBe(2);
    expect(GRANT_UNIT_ROWS.filter((r) => r.tech === 'POTTERY').length).toBe(1);
    const born = (row: number): number => {
      const state = sceneAs(row);
      const city = settleAt(state, tileAtCoords(state.map, 7, 7).index, 0);
      state.seats[0].capitalTile = city.centerIndex;
      state.seats[0].research.tech = 'POTTERY';
      state.seats[0].research.techProgress = TECHS.POTTERY.cost * 10;
      state.unitsMode = true;
      seatPhase(state);
      return state.units.filter((u: { type: string; seat: number }) => u.type === 'TRADER' && u.seat === 0).length;
    };
    expect(born(seatRow('CREE'))).toBe(1);
    expect(born(seatRow('AMERICA'))).toBe(0);
  });

  it("Catherine's Spy and her capacity at Castles", () => {
    const state = sceneAs(leaderRow('CATHERINE_DE_MEDICI'));
    settleAt(state, tileAtCoords(state.map, 7, 7).index, 0);
    const base = spyCapacity(state, 0);
    grantTechs(state, 'CASTLES');
    expect(spyCapacity(state, 0)).toBe(base + 1);
    const plain = sceneAs(seatRow('AMERICA'));
    settleAt(plain, tileAtCoords(plain.map, 7, 7).index, 0);
    const pb = spyCapacity(plain, 0);
    grantTechs(plain, 'CASTLES');
    expect(spyCapacity(plain, 0)).toBe(pb);
  });

  it("Kupe's Voyage: the first city starts at 2 Population with a Builder, and the Palace pays more", () => {
    expect(CAPITAL_ROWS.length).toBe(1);
    const state = sceneAs(leaderRow('KUPE'));
    state.unitsMode = true;
    const city = foundCityAt(state, 0, tileAtCoords(state.map, 7, 7), state.seats[0]);
    expect(city.population).toBe(2);
    expect(state.units.filter((u) => u.type === 'BUILDER').length).toBe(1);
    const second = foundCityAt(state, 0, tileAtCoords(state.map, 3, 3), state.seats[0]);
    expect(second.population).toBe(1);
    expect(state.units.filter((u) => u.type === 'BUILDER').length).toBe(1);
    const plain = sceneAs(seatRow('AMERICA'));
    const pc = foundCityAt(plain, 0, tileAtCoords(plain.map, 7, 7), plain.seats[0]);
    expect(pc.population).toBe(1);
    // the Palace's housing and amenity, over the plain seat's
    const kupeH = computeCityStats(state, city).housing;
    const plainH = computeCityStats(plain, pc).housing;
    expect(kupeH).toBe(plainH + 3);
  });

  it('+2 Science and +2 Culture per turn before the first city', () => {
    const state = sceneAs(leaderRow('KUPE'));
    const sci = state.seats[0].research.techProgress;
    spawnUnit(state, 'SETTLER', tileAtCoords(state.map, 7, 7).index, 0);
    seatPhase(state);
    expect(state.seats[0].research.techProgress).toBe(sci + 2);
    expect(state.seats[0].research.civicProgress).toBe(2);
  });
});

File: tests/gpu/city_rows_test.py
"""THE CITY'S ROSTER ROWS — the GPU half.

    npm run seed && npm run export        # (once) writes seeder/worlds/
    python tests/gpu/city_rows_test.py

The TS twin is tests/cpu/seats/city-rows.test.ts.

CIV6 (the install's TraitModifiers): the centre's terrain adjacency (Songs
of the Jeli), the per-work yields and the Great Person factor (Nkisi), the
powered building's extra yield, strategic accumulation and its ceiling,
build charges (Workshop of the World), the tile price and the Farm's
ground (The Last Best West), the route's per-improvement yields (Favorable
Terms), the granted units and spy capacity, and the first city (Kupe).
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

import torch

sys.path.insert(0, str(Path(__file__).resolve().parent.parent.parent / "gpu"))
sys.path.insert(0, str(Path(__file__).resolve().parent))
from core import BatchSim, load_rules, load_fixture, fixture_paths
from warmup import settle_all

B0 = 0
RULES = json.loads((Path(__file__).resolve().parent.parent.parent
                    / "seeder" / "worlds" / "rules.json").read_text())
UNITS = [u["id"] for u in RULES["units"]]
BLDGS = [b["id"] for b in RULES["buildings"]]
TECHS = [t["id"] for t in RULES["techs"]]
IMPS = RULES["improvements"]["ids"]


def play(sim, row: int, name):
    if name is None:
        sim.row_civ[0, row] = -1
        sim.row_leader[0, row] = -1
    else:
        ci = sim._civ_ids.index(name)
        sim.row_civ[0, row] = ci
        sim.row_leader[0, row] = sim._pair_civ.index(ci)
    sim._eff_version += 1
    sim._gen_ver += 1
    sim._bldg_version += 1


def fresh(rules, path) -> BatchSim:
    sim = BatchSim([load_fixture(path)], rules, device="cpu", dtype=torch.float64)
    for r, name in enumerate(("ROME", "EGYPT", "NORWAY")):
        play(sim, r, name)
    return settle_all(sim)


def bare(rules, path) -> BatchSim:
    """Unsettled, so a founding clause can be measured at its own moment."""
    sim = BatchSim([load_fixture(path)], rules, device="cpu", dtype=torch.float64)
    for r in range(sim.n_majors):
        play(sim, r, None)
    return sim


def T(*xs) -> torch.Tensor:
    return torch.tensor(list(xs), dtype=torch.long)


# ---------------------------------------------------------------------------


def test_wire(rules, path) -> None:
    sim = fresh(rules, path)
    assert len(sim._center_adj_rows) == 2 and len(sim._great_work_yield_rows) == 8
    assert len(sim._gpp_class_rows) == 3 and len(sim._powered_yield_rows) == 5
    assert len(sim._stockpile_rate_rows) == 4 and len(sim._stockpile_cap_rows) == 3
    # the charge family grew with Qin's Builder, Philip's Inquisitor and
    # Dharma's Missionary — a census pin is a count, so it moves with them
    assert len(sim._unit_charge_rows) == 4 and len(sim._tile_cost_rows) == 2
    assert len(sim._farm_terrain_rows) == 2 and len(sim._route_improvement_rows) == 4
    # the TECH-keyed grant rows, not the list's length: the family also
    # carries founding grants now (`foreignContinent`)
    assert sum(1 for r in sim._grant_unit_rows if r[3] >= 0) == 2
    assert len(sim._spy_capacity_rows) == 1
    assert len(sim._capital_rows) == 1
    print("  1 wire OK — the city's thirteen row families")


def test_district_production(rules, path) -> None:
    """A district queue code is a SCAFFOLD slot, not a district index: the row
    names the district, so only that district's own slot may pay."""
    sim = fresh(rules, path)
    dcat = [d["id"] for d in sim.districts_cat]

    def slot_of(name: str) -> int:
        di = dcat.index(name)
        return next(si for si, p in enumerate(sim._scaffold) if p[0] == di)

    def produce(name, district: str) -> float:
        s2 = fresh(rules, path)
        play(s2, 0, name)
        s2.city_current[B0, 0, 0, 0] = s2.DISTRICT_BASE + slot_of(district)
        s2.city_qtile[B0, 0, 0, 0] = int(s2.city_center[B0, 0, 0])
        s2.city_progress[B0, 0, 0, 0] = 0
        s2.city_cost[B0, 0, 0, 0] = 100000
        s2._eff_version += 1
        s2._seat_city_produce(0, torch.tensor([0]), torch.tensor([True]),
                              torch.tensor([20.0], dtype=torch.float64))
        return float(s2.city_progress[B0, 0, 0, 0])

    base = produce("ROME", "DAM")
    assert base > 0
    assert abs(produce("NETHERLANDS", "DAM") - base * 1.5) < 1e-9, "Grote Rivieren's Dam"
    assert abs(produce("NETHERLANDS", "PRESERVE") - produce("ROME", "PRESERVE")) < 1e-9, "the Dam row paid a Preserve"
    assert abs(produce("JAPAN", "HOLY_SITE") - produce("ROME", "HOLY_SITE") * 2.0) < 1e-9, "Divine Wind's half time"
    assert abs(produce("JAPAN", "CAMPUS") - produce("ROME", "CAMPUS")) < 1e-9, "Divine Wind paid a Campus"
    print("  2 district production OK — the Dam, the three of Divine Wind, and nothing else")


def test_powered_and_gpp(rules, path) -> None:
    sim = fresh(rules, path)
    play(sim, 0, "ENGLAND")
    add = sim._powered_add(0)
    assert add is not None and float(add[B0, 3]) == 4.0 and float(add[B0, 2]) == 4.0
    play(sim, 0, "ROME")
    assert sim._powered_add(0) is None or float(sim._powered_add(0)[B0].abs().sum()) == 0.0
    play(sim, 0, "KONGO")
    artist = next(i for i, r in enumerate(sim._gpp_class_rows) if r[2] >= 0)
    cls = sim._gpp_class_rows[artist][2]
    assert abs(float(sim._gpp_class_mult(0, cls)[B0]) - 1.5) < 1e-12
    other = next(c for c in range(sim.civ_gpp.shape[2]) if c not in {r[2] for r in sim._gpp_class_rows})
    assert float(sim._gpp_class_mult(0, other)[B0]) == 1.0
    play(sim, 0, "ROME")
    assert float(sim._gpp_class_mult(0, cls)[B0]) == 1.0
    print("  2 powered yields + the Great Person factor OK")


def test_stockpile_and_charges(rules, path) -> None:
    sim = fresh(rules, path)
    play(sim, 0, "ENGLAND")
    base = int(sim._stockpile_cap(0)[B0])
    sim.city_bldg[B0, 0, 0, BLDGS.index("LIGHTHOUSE")] = True
    assert int(sim._stockpile_cap(0)[B0]) == base + 10, "Workshop of the World's ceiling"
    sim.city_bldg[B0, 0, 0, BLDGS.index("SEAPORT")] = True
    assert int(sim._stockpile_cap(0)[B0]) == base + 20
    play(sim, 0, "ROME")
    assert int(sim._stockpile_cap(0)[B0]) == base, "the ceiling outlived England"
    me = UNITS.index("MILITARY_ENGINEER")
    play(sim, 0, "ENGLAND")
    assert int(sim._extra_charges(0, T(me))[B0]) == 2
    assert int(sim._extra_charges(0, T(UNITS.index("BUILDER")))[B0]) == 0
    play(sim, 0, "ROME")
    assert int(sim._extra_charges(0, T(me))[B0]) == 0
    print("  3 stockpile ceiling + build charges OK")


def test_stockpile_rate(rules, path) -> None:
    """Each resource row of `STOCKPILE_RATE_ROWS`, on a mine of its own kind."""
    def bank_of(name, rid: int) -> int:
        sim = fresh(rules, path)
        play(sim, 0, name)
        k = next(i for i, r in enumerate(sim._strat_rid) if int(r) == rid)
        mine = IMPS.index("MINE")
        t = next(t for t in range(sim.T)
                 if int(sim.tile_seat[B0, t]) == 0 and not bool(sim.water[B0, t]))
        sim.res_id[B0, t] = rid
        sim.res_imp[B0, t] = mine
        sim.improvement[B0, t] = mine
        sim.pillaged[B0, t] = False
        sim.civ_stockpile[B0, 0, :] = 0
        sim._seat_accrue_stockpile(0)
        return int(sim.civ_stockpile[B0, 0, k])

    sim = fresh(rules, path)
    rows = [r for r in sim._stockpile_rate_rows if r[2] >= 0]
    assert rows, "no resource rate row on the wire"
    for _sc, _sl, rid, _st, amt, _pct in rows:
        assert bank_of("ENGLAND", rid) == bank_of("ROME", rid) + amt, f"resource {rid} accumulates {amt} more"
    print(f"  4 accumulation rows OK — {len(rows)} resources at +2 per mine")


def test_tile_price(rules, path) -> None:
    sim = fresh(rules, path)
    tundra = next((t for t in range(sim.T) if int(sim.terrain[B0, t]) == 3), None)
    if tundra is None:
        print("  5 tile price SKIPPED — this fixture holds no Tundra")
        return
    ctr = sim.city_center[B0, 0, 0].reshape(1)
    tgt = torch.tensor([tundra], dtype=torch.long)
    play(sim, 0, "ROME")
    plain = float(sim._seat_tile_price(0, ctr, tgt)[B0])
    play(sim, 0, "CANADA")
    assert abs(float(sim._seat_tile_price(0, ctr, tgt)[B0]) - round(plain * 0.5)) < 1e-9, "The Last Best West"
    print("  5 tile price OK — half on the tundra")


def build_at(sim, t: int, k: int, row: int = 0) -> bool:
    """Spawn a Builder ON tile `t` and issue BUILD_<k> over the real order
    path (the geothermal lane's helper). Did the improvement land?"""
    slot = int(sim.unit_next[B0])
    sim._spawn_unit(row, torch.ones(sim.B, dtype=torch.bool),
                    torch.full((sim.B,), int(sim.city_center[B0, row, 0]), dtype=torch.long),
                    sim._builder_idx)
    old_t = int(sim.unit_tile[B0, slot])
    rows = torch.tensor([B0])
    sim._occ_clear(rows, torch.tensor([old_t]), torch.tensor([slot]))
    sim.unit_tile[B0, slot] = t
    sim._occ_set(rows, torch.tensor([t]), torch.tensor([slot]))
    sim.unit_mp[B0, slot] = sim._mp_scale
    smap = sim._seat_slot_map(row)
    rank = int((smap[B0] == slot).nonzero()[0])
    act = torch.full(smap.shape, -1, dtype=torch.long)
    act[B0, rank] = sim._A_IMP[k]
    sim._apply_seat_unit_actions(row, act)
    return int(sim.improvement[B0, t]) == k


def tundra_plot(sim, hills: bool) -> int | None:
    """An owned, empty land tile made Tundra — the row's own ground."""
    t = next((t for t in range(sim.T)
              if int(sim.tile_seat[B0, t]) == 0 and not bool(sim.water[B0, t])
              and bool(sim.passable[B0, t]) and int(sim.improvement[B0, t]) < 0
              and int(sim.district[B0, t]) < 0 and int(sim.centre_slot_at[B0, t]) < 0
              and int(sim.built_wonder[B0, t]) < 0), None)
    if t is None:
        return None
    sim.terrain[B0, t] = 3  # `TERRAIN_IDS`[3] is TUNDRA
    sim.hills[B0, t] = hills
    sim.feat_id[B0, t] = -1
    sim.res_id[B0, t] = -1
    sim.res_cat[B0, t] = 0
    sim.farm_flat[B0, t] = False
    sim.farm_hill[B0, t] = False
    sim._eff_version += 1
    return t


def test_farm_ground(rules, path) -> None:
    farm = IMPS.index("FARM")
    civil = next((i for i, c in enumerate(RULES["civics"]) if c["id"] == "CIVIL_ENGINEERING"), -1)

    def farms(name, hills: bool, civic: bool) -> bool:
        sim = fresh(rules, path)
        play(sim, 0, name)
        for tname in ("POTTERY", "IRRIGATION"):
            sim.civ_techs[B0, 0, TECHS.index(tname)] = True
        if civic and civil >= 0:
            sim.civ_civics[B0, 0, civil] = True
        sim._eff_version += 1
        t = tundra_plot(sim, hills)
        assert t is not None, "no owned empty land tile"
        return build_at(sim, t, farm)

    assert farms("CANADA", False, False), "The Last Best West did not farm the Tundra"
    assert not farms("ROME", False, False), "a plain seat farmed the Tundra"
    assert not farms("CANADA", True, False), "Tundra Hills before Civil Engineering"
    assert farms("CANADA", True, True), "Tundra Hills at Civil Engineering"
    print("  6 farm ground OK — Tundra, and its hills at Civil Engineering")


def test_grants_and_capital(rules, path) -> None:
    sim = bare(rules, path)
    play(sim, 0, "MAORI")
    n0 = int(sim.unit_next[B0])
    settle_all(sim)
    assert int(sim.city_pop[B0, 0, 0]) == 2, "Kupe's first city starts at 2 Population"
    builder = UNITS.index("BUILDER")
    made = [s for s in range(n0, int(sim.unit_next[B0]))
            if bool(sim.unit_alive[B0, s]) and int(sim.unit_type[B0, s]) == builder
            and int(sim.unit_seat[B0, s]) == 0]
    assert len(made) == 1, "Kupe's free Builder"
    plain = bare(rules, path)
    play(plain, 0, "ROME")
    p0 = int(plain.unit_next[B0])
    settle_all(plain)
    assert int(plain.city_pop[B0, 0, 0]) == 1
    assert not [s for s in range(p0, int(plain.unit_next[B0]))
                if bool(plain.unit_alive[B0, s]) and int(plain.unit_type[B0, s]) == builder
                and int(plain.unit_seat[B0, s]) == 0], "the Builder outlived Kupe"
    print("  7 Kupe's first city OK — 2 Population and a Builder")


def test_spy_capacity(rules, path) -> None:
    sim = fresh(rules, path)
    castles = TECHS.index("CASTLES")
    play(sim, 0, "FRANCE")
    if int(sim.row_leader[B0, 0]) != sim._leader_idx("CATHERINE_DE_MEDICI"):
        # France's first row is Catherine; a later row would be Eleanor
        sim.row_leader[B0, 0] = sim._leader_idx("CATHERINE_DE_MEDICI")
        sim._eff_version += 1
    sim.civ_techs[B0, 0, castles] = False
    base = int(sim._spy_capacity(0)[B0])
    sim.civ_techs[B0, 0, castles] = True
    assert int(sim._spy_capacity(0)[B0]) == base + 1, "the Flying Squadron"
    play(sim, 0, "ROME")
    assert int(sim._spy_capacity(0)[B0]) == base, "the capacity outlived Catherine"
    print("  8 spy capacity OK — +1 at Castles")


def main() -> int:
    rules = load_rules()
    path = fixture_paths()[0]
    test_wire(rules, path)
    test_district_production(rules, path)
    test_powered_and_gpp(rules, path)
    test_stockpile_and_charges(rules, path)
    test_stockpile_rate(rules, path)
    test_tile_price(rules, path)
    test_farm_ground(rules, path)
    test_grants_and_capital(rules, path)
    test_spy_capacity(rules, path)
    print("BATTERY OK city_rows")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
